File: 2025/day07_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    return propagate(
        find_start(content[0]),
        list(
            enumerate(
                [line for line in content[1:] if any(ch != "." for ch in line)],
            )
        ),
    )


cache: dict[tuple[int, int], int] = {}


def propagate(beam: int, content: list[tuple[int, str]]) -> int:
    if not content:
        return 1
    (row, line), *rest = content
    key = (row, beam)
    cached = cache.get(key)
    if cached is not None:
        return cached
    if line[beam] == "^":
        value = propagate(beam - 1, rest) + propagate(beam + 1, rest)
        cache[key] = value
        return value
    return propagate(beam, rest)


def find_start(line: str) -> int:
    return line.index("S")


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2025/day07-2.hs
import Data.List (elemIndex)
import Data.Map (Map, empty, insert, lookup, member)
import Data.Maybe (fromJust)
import Prelude hiding (lookup)

type Cache = Map (Int, Int) Int

solution :: String -> Int
solution content = fst . propagate empty (findStart . head . lines $ content) $ zip [0 ..] (filter (any (/= '.')) . tail . lines $ content)

propagate :: Cache -> Int -> [(Int, String)] -> (Int, Cache)
propagate cache beam [] = (1, cache)
propagate cache beam ((row, line) : xs)
  | member key cache = (fromJust $ lookup key cache, cache)
  | (!! beam) line == '^' = (value, insert key value nextCache)
  | otherwise = propagate cache beam xs
  where
    key = (row, beam)
    (prev, prevCache) = propagate cache (beam - 1) xs
    (next, nextCache) = propagate prevCache (beam + 1) xs
    value = prev + next

findStart :: String -> Int
findStart = fromJust . elemIndex 'S'

main :: IO ()
main = do readFile "day07.dat" >>= print . solution
